checker/src/org/checkerframework/checker/lock/LockVisitor.java

```diff
@@ -133,166 +411,116 @@ protected boolean checkOverride(MethodTree overriderTree,
+                    receiverIsThatOfEnclosingMethod = true;
+                }
+            }
 
-                                        additionalPreconditions.add(Pair.of(lockExpression, checkerLockHeldClass.toString().substring(10 /* "interface " */)));
-                                    }
-                                }
-                            }
+            if (expr != null && atmOfReceiverOrIdentifier != null) {
+                AnnotationMirror gb = atmOfReceiverOrIdentifier.getEffectiveAnnotationInHierarchy(GUARDEDBYUNKNOWN);
+                // IMPORTANT: This relies on getEffectiveAnnotationInHierarchy being sound in the sense that
+                // it never returns null if an effective annotation in the @GuardedByUnknown hierarchy was present.
+                // It is critical to the soundness of the Lock Checker that an effective primary @GuardedBy(...) annotation
+                // never go unnoticed when checking the access of an expression.
+                if (gb != null) {
+                    if (AnnotationUtils.areSameByClass( gb, checkerGuardedByClass)) {
+                        // It is critical that if receiverIsThatOfEnclosingMethod is true,
+                        // generatePreconditionsBasedOnGuards translate the expression
+                        // "itself" to "this". That's because right now we know that, since
+                        // we are dealing with the receiver of the method, "itself" corresponds
+                        // to "this". However once checkPreconditions is called, that
+                        // knowledge is lost and it will regards "itself" as referring to
+                        // the variable the precondition we are about to add is attached to.
+
+                        if (invokedElement != null) {
+                            checkPreconditions(expr, invokedElement, expr.getKind() == Tree.Kind.METHOD_INVOCATION,
+                                generatePreconditionsBasedOnGuards(atmOfReceiverOrIdentifier,
+                                        receiverIsThatOfEnclosingMethod /* see comment above */),
+                                        false);
                         }
+                    } else if (AnnotationUtils.areSameByClass(gb, checkerGuardSatisfiedClass)){
+                        // Can always dereference if type is @GuardSatisfied
+                    } else {
+                        // Can never dereference for any other types in the @GuardedBy hierarchy
+                        String annotationName = gb.toString();
+                        annotationName = annotationName.substring(annotationName.lastIndexOf('.') + 1 /* +1 to skip the last . as well */);
+                        checker.report(Result.failure(
+                                "cannot.dereference",
+                                tree.toString(),
+                                "annotation @" + annotationName), tree);
+                        return;
```

Comment:
It is for now, but my intent in adding it was to make sure any code added after this point in the method would not be executed in this case. However that much future-proofing should not be needed in this case. I'll remove it.